### ISSUE TITLE
Current Step about differential model, and what's the next

### ISSUE THREAD

**Issue by cd155** [JacquesCarette/Drasil#2898]
Here is the current stage of "Adding ODE knowledge in Drasil", after completing #2884, 

1. Created a new data type has different ODE data represent.
```haskell
data DifferentialModel = Linear {
                                  _indepVar :: UnitalChunk,
                                  _depVar :: ConstrConcept,
                                  _coefficients :: [CoeffDeriv],
                                  _constant :: Expr,
                                  _conc :: ConceptChunk
                                }
```
The default method will display canonical form. We also like to add new functions to display ODE to other forms. Here is the related ticket for adding a display form with constant on the rhs. #2877 

2. Examples changed to NewDEModel
    - [x] PDController (completed)
    - [x] NoPCM (completed)

3. Example not ready 
    - [ ] SWHS (not completed)
         - the SWHS has three cases on the rhs of its ODE. Base on the Dr.Smith comment, they called ODEs with Events. Here is the related ticket about adding ODEs with Events in Drasil. #2878
         
### Project Goals Break down
Here is a brief summary of the goals we want to accomplish based on previous discussions.

Big Goal
- Step 1: Teach Drasil about ODE
- Step 2: How to hook up an ODE solver to an ODE
- Step 3: Option for remaining situations

Smaller Goal for Adding NewDEModel:
- [x] Step 1: change current DEModel to NewDEModel
- [ ] Step 2: take other unready differential equation models into NewDEModel (**I actually not sure if I understand this step right**)
- [ ] Step 3: the meaning of derivative on the lhs. Where are they? what do they mean? where does the differential model belong?

In summary, the basic template of NewDEModel is completed, so I wonder to know which direction should I go in the next couple of weeks.

**Comment by smiths** [comment]
I have a few thoughts that will need feedback from @JacquesCarette.

1. Connect the NewDEModel to the code generator.
2. Generate a new example (SRS and code) for a simple linear ODE solver.  The goal would not be to document the full family of ODE solvers, just one example.  We could think of it as a document to teach math students about ODEs.  The inputs would be the coefficients for the linear ode and the initial conditions.
3. We could add another physics example, or more than one, with some additional odes cases that don't currently exist in the examples.  (I feel like this is too much physics and not enough ODE.)
4. Start brainstorming on how to understand the family of ODE models.  What are the commonalities and what are the variabilities?  We should stay away from computational variabilities (algorithm variabilities) for now.

**Comment by JacquesCarette** [comment]
About @smiths 's few thoughts: these are all good things, but at very different scales. Your 1 belongs to this task, but I don't think that 2-4 necessarily do.

About @cd155 's question on Step 2: You need to find all differential equations in all (current) models. Then we need to understand what prevents the ones that are not in NewDEModel, why not. This could be an empty step, I don't know!

For Step 3, the problem might be quite different. The issue might not be one of modelling at all, but one of computation. In other words, the specification (say of acceleration as the derivative to velocity with respect to time) is quite well-defined, but this does not give us a real means to compute accelaration (numerical differentiation is extremely unstable, and should be seen as a non-solution). So we need to understand more deeply what we're really saying when we specify things that way.

**Comment by cd155** [comment]
Thanks for the feedback.

@smiths For the next couple of weeks, I will start working on connecting the `NewDEModel` to the code generator. In the summer, I tried manually creating an `ODEInfo` which contains ODEs of the double pendulum example. I didn't complete it at the time, but I think I should be able to start to finish it now.

@JacquesCarette  For now, only the example of swhs still uses the previous `deModel` model. The reason it hasn't been changing to `NewDEModel` is because there are multiple equations on the right-hand side. Meanwhile, I think the single pendulum and double pendulum example should include `NewDEModel`, and I created an issue #2901 to discuss them. 
Last, for the Step 3 

> For Step 3, the problem might be quite different. The issue might not be one of modeling at all, but one of computation. In other words, the specification (say of acceleration as the derivative to velocity with respect to time) is quite well-defined, but this does not give us a real means to compute acceleration (numerical differentiation is extremely unstable, and should be seen as a non-solution). So we need to understand more deeply what we're really saying when we specify things that way.

I am not quite sure I understand the guideline. Is it possible to be more specific? To take the acceleration as the example, it can be written as 

<img width="100" alt="Screenshot from 2021-11-29 15-36-53" src="https://user-images.githubusercontent.com/16947266/143939353-69667f6d-304b-4999-9cca-b0be1518f7f4.png">

To compute the acceleration, we can either use (the derivative to velocity with respect to time) or (the second derivative to distance with respect to time), why this didn't give a real meaning? Sorry if I ask a bad question, I just feel I have some disconnections to understand Step 3.

**Comment by JacquesCarette** [comment]
Is there an issue to discuss the multiple right-hand-sides of SWHS?

Your question on acceleration is excellent. Mathematically, using either of those definitions (derivative of velocity or 2nd derivative of position) is perfect. Computationally however, computing derivatives numerically is hard and not a good idea. So either we can compute those derivatives symbolically (and right now, that would have to be done by the person writing the specification), or we're stuck. In the language of Drasil, using the above as a DataDefinition that would be expected to be translated 'literally' to code would fail.

Note that there is a subtle point: specifying acceleration that way is not a *differential equation* because there is nothing to 'solve for'. Mathematically, it is just a plain definition. It just turns out to be a definition that we can't use directly in code. As it isn't an issue of *models* but one of *definitions*, it is a bit to the side of `newDEModel`.

**Comment by cd155** [comment]
@JacquesCarette  Here is the ticket about multiple right-hand-sides of SWHS #2878

Thanks for explaining it again, I think I have a better understanding of it now.
